### PR TITLE
Company register UIformValidation

### PR DESCRIPTION
**Pull request by TBalo**

### CODE REVIEW

File: components/CompanyRegister.tsx [hunk 1]
```diff
@@ -1,42 +1,72 @@
...
+import { useRecruitmentRegisterCompany } from '../src/generated';
 import { useRouter } from 'next/router';
 import { useEffect, useState } from 'react';
 import { toast } from 'react-toastify';
-import { useAccount } from 'wagmi';
-import { useRecruitmentRegisterCompany } from '../src/generated';
+import { isSuccessResponse } from '@/pages/api/companies';
 import { Badge } from './ui/badge';
 
+const schema = yup.object().shape({
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: lets switch to zod.  It has better type safety:

https://www.npmjs.com/package/@hookform/resolvers#zod

File: components/CompanyRegister.tsx [hunk 2]
```diff
@@ -1,42 +1,72 @@
-import { CreateCompanyItem } from '@/db/entities/company';
+import { yupResolver } from '@hookform/resolvers/yup';
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: organise imports with Shift+Alt+O

File: components/CompanyRegister.tsx [hunk 3]
```diff
@@ -1,42 +1,72 @@
...
+interface FormData {
+  companyName: string;
+  companyEmail: string;
+  companySite: string;
+}
+
 const CompanyRegister = () => {
   const router = useRouter();
   const { address } = useAccount();
-  const { isValidating, isRegistered } = useCompany(address);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx:14]: form submit button should be disabled until `isValidating`=== false.

So this needs to be added back in

File: components/CompanyRegister.tsx [hunk 4]
```diff
@@ -1,42 +1,72 @@
...
-  const [companyName, setCompanyName] = useState('');
-  const [companyEmail, setCompanyEmail] = useState('');
-  const [companySite, setCompanySite] = useState('');
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    getValues,
+  } = useForm<FormData>({
+    resolver: yupResolver(schema) as any,
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: we want to use `any` as little as possible in our code

```suggestion
    resolver: yupResolver(schema),
```

File: components/CompanyRegister.tsx [hunk 5]
```diff
@@ -1,42 +1,72 @@
...
+  });
 
-  const registerCompanySc = async () => {
-    if (isRegistered) {
+  const onSubmit = async (data: FormData) => {
+    if (!isRegistered) {
+      try {
+        await writeAsync();
+        registerCompanyDb(data); // Pass the data received from the form
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: this is handled by the useEffect later
```suggestion
 
```

File: components/CompanyRegister.tsx [hunk 6]
```diff
@@ -1,42 +1,72 @@
...
-      await writeAsync();
-    } catch (e) {
-      toast.error('Company registration failed. Please contact support');
     }
   };
 
-  const registerCompanyDb = async () => {
-    const companyData: CreateCompanyItem = {
+  const registerCompanyDb = async (data: FormData) => {
+    const companyData = {
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: We want to use `CreateCompanyItem` as this is what the API we are about to post to expects
```suggestion
    const companyData: CreateCompanyItem = {
```

File: components/CompanyRegister.tsx [hunk 7]
```diff
@@ -49,80 +79,65 @@ const CompanyRegister = () => {
...
       if (response.ok) {
         const responseData = await response.json();
-
         if (isSuccessResponse(responseData)) {
           toast.success(`${responseData.item.name} Registered Successfully`);
           router.push('/');
+        } else {
+          toast.error('Registration Unsuccessful. Please contact support.');
         }
-      } else {
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx:57]: if response.ok is false we want to know about it still - lets keep this else

File: components/CompanyRegister.tsx [hunk 8]
```diff
@@ -49,80 +79,65 @@ const CompanyRegister = () => {
...
-  const handleCompanySiteChange = (event: any) => {
-    setCompanySite(event.target.value);
-  };
 
   useEffect(() => {
     if (isSuccess) {
-      registerCompanyDb();
-    }
+      const formData = getValues();
+      registerCompanyDb(formData);    }
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: formatting of the brackets

**Review by codingnuclei — CHANGES_REQUESTED** [review]
Thanks for contributing 💯 

A good start. A few comments with room for improvements

File: components/CompanyRegister.tsx (hunk 1 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: noted, zod dependency has been installed and used

File: components/CompanyRegister.tsx (hunk 2 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx (hunk 3 above)
Comment (TBalo) [reply components/CompanyRegister.tsx:14]: re-added

File: components/CompanyRegister.tsx (hunk 4 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx (hunk 5 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx (hunk 6 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: was added

File: components/CompanyRegister.tsx (hunk 7 above)
Comment (TBalo) [reply components/CompanyRegister.tsx:57]: done. was actually added earlier missed that.

File: components/CompanyRegister.tsx (hunk 8 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx [hunk 9]
```diff
@@ -1,128 +1,145 @@
...
+  companyEmail: string;
+  companySite: string;
+}
+
 const CompanyRegister = () => {
   const router = useRouter();
   const { address } = useAccount();
-  const { isValidating, isRegistered } = useCompany(address);
+  const { isRegistered } = useCompany(address);
+  const [isValidating] = useState(false);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: opps - isValidating comes from the useCompany hook :)

File: components/CompanyRegister.tsx [hunk 10]
```diff
@@ -1,128 +1,145 @@
...
+    getValues,
+  } = useForm<FormData>({ resolver: zodResolver(schema) });
 
-  const registerCompanySc = async () => {
-    if (isRegistered) {
+  const onSubmit = async (data: FormData) => {
+    if (!isRegistered) {
+      try {
+        await writeAsync();
+        registerCompanyDb(data);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: missed this line again.
```suggestion

```

it is handled in the useEffect

File: components/CompanyRegister.tsx [hunk 11]
```diff
@@ -1,128 +1,145 @@
...
-
-  const handleCompanyEmailChange = (event: any) => {
-    setCompanyEmail(event.target.value);
-  };
-
-  const handleCompanySiteChange = (event: any) => {
-    setCompanySite(event.target.value);
-  };
-
-  useEffect(() => {
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx:78]: why has this been removed - it is required

File: package.json [hunk 12]
```diff
@@ -48,7 +50,9 @@
     "ulidx": "^2.2.1",
     "uuid": "^9.0.1",
     "viem": "^1.18.0",
-    "wagmi": "^1.4.5"
+    "wagmi": "^1.4.5",
+    "yup": "^1.3.2",
```
Comment (codingnuclei) [inline package.json]: remove yup

File: components/CompanyRegister.tsx [hunk 13]
```diff
@@ -1,128 +1,145 @@
...
-    };
-    const options = {
-      method: 'POST',
-      headers: {
-        'Content-Type': 'application/json',
-      },
-      body: JSON.stringify(companyData),
-    };
+  const registerCompanyDb = async (data: FormData) => {
+    const validatedData = schema.safeParse(data);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: the form can not be submitted unless valid so is this an unneeded check?

File: components/CompanyRegister.tsx [hunk 14]
```diff
@@ -1,128 +1,145 @@
...
 
+const schema = z.object({
+  companyName: z
+    .string()
+    .min(1, { message: 'Company Name is required' })
+    .max(255),
+  companyEmail: z.string().email({ message: 'Invalid email format' }),
+});
+
+interface FormData {
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]: do we need this type as we are using a zod schema?

**Review by codingnuclei — CHANGES_REQUESTED** [review]

File: components/CompanyRegister.tsx (hunk 9 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: Re-adjusted

File: components/CompanyRegister.tsx (hunk 10 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: i left it and removed the useEffect as what it's just doing is to call registerDb function when the state of the form changes. But this appears to be a duplicated action

File: components/CompanyRegister.tsx (hunk 11 above)
Comment (TBalo) [reply components/CompanyRegister.tsx:78]: it was added to call registerDb function when the state of the form changes. But this appears to be a duplicated action

File: package.json (hunk 12 above)
Comment (TBalo) [reply package.json]: removed

File: components/CompanyRegister.tsx (hunk 13 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: yeah truly adjusted.

File: components/CompanyRegister.tsx (hunk 3 above)
Comment (codingnuclei) [reply components/CompanyRegister.tsx:14]: `isValidating` is from the `useCompany` hook.  The hook checks if the company is already registered.  If the company is registered then the register button should be disabled. `isValidating` is true while the hook is checking if the company is registered.  We want to also disable the register button while checking if the company is registered.

File: components/CompanyRegister.tsx [hunk 15]
```diff
@@ -1,128 +1,142 @@
...
+
+type FormData = z.infer<typeof schema>;
+
 
 const CompanyRegister = () => {
   const router = useRouter();
   const { address } = useAccount();
-  const { isValidating, isRegistered } = useCompany(address);
+  const { isRegistered } = useCompany(address);
+  const [isValidating, setIsValidating] = useState(false);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]:
```suggestion

```

File: components/CompanyRegister.tsx (hunk 10 above)
Comment (codingnuclei) [reply components/CompanyRegister.tsx]: we want your PR to be focussed purely on form validation.

By adding registerCompanyDb here you have changed the pattern used to trigger a DB save after a SC write.  That change is not in scope of your PR...

Please revert to the useEffect pattern that was already in place

File: components/CompanyRegister.tsx (hunk 11 above)
Comment (codingnuclei) [reply components/CompanyRegister.tsx:78]: this useEffect was here to save to DB once SC had been written (`isSuccess`).

Please stay focussed on changing form validation only!!

File: yarn.lock [hunk 16]
```diff
@@ -9264,6 +9294,16 @@ yocto-queue@^1.0.0:
   resolved "https://registry.npmjs.org/yocto-queue/-/yocto-queue-1.0.0.tgz"
   integrity sha512-9bnSc/HEW2uRy67wc+T8UwauLuPJVn28jb+GtJY16iiKWyvmYJRXVT4UamsAEGQfPohgr2q4Tq0sQbQlxTfi1g==
 
+yup@^1.3.2:
```
Comment (codingnuclei) [inline yarn.lock]: bit weird that yup is still in the lockfile? can you try a `yarn install` and commit the latest lockfile that is output.

**Review by codingnuclei — CHANGES_REQUESTED** [review]
Some more feedback.

Really need it PR to be focussed on the task at hand and not changing other aspects of the file.  The more things that have been changed the bigger the risk of merging your PR

File: yarn.lock (hunk 16 above)
Comment (TBalo) [reply yarn.lock]: fixed

File: components/CompanyRegister.tsx (hunk 15 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx [hunk 17]
```diff
@@ -1,128 +1,144 @@
...
+    register,
+    handleSubmit,
+    getValues,
+    formState: { errors, isValid },
+  } = useForm<FormData>({ resolver: zodResolver(schema) });
 
-  const registerCompanySc = async () => {
-    if (isRegistered) {
+  const onSubmit: SubmitHandler<FormData> = async (data: FormData) => {
+    if (!isRegistered) {
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]:
```suggestion
    if (!isRegistered && isValid) {
```

File: components/CompanyRegister.tsx [hunk 18]
```diff
@@ -1,128 +1,144 @@
...
+              className='h-[40px] w-[200px] rounded-lg border border-slate-500 md:w-[20vw]'
+            />
+            {errors.companySite && typeof errors.companySite === 'string' && (
+              <span className='text-red-500'>{errors.companySite}</span>
+            )}
+          </div>
+          <button
+            type='submit'
+            className='btn btn-primary mt-5 w-[100%]'
+            disabled={isLoading || isValidating || !isValid}
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx]:
```suggestion
            disabled={isLoading || isValidating}
```

**Review by codingnuclei — CHANGES_REQUESTED** [review]

File: components/CompanyRegister.tsx (hunk 17 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

File: components/CompanyRegister.tsx (hunk 18 above)
Comment (TBalo) [reply components/CompanyRegister.tsx]: done

**Review by codingnuclei — APPROVED** [review]

**Pull request merged**